### PR TITLE
feat: move tasks to workers

### PR DESCRIPTION
**Pull request by NoemiRozpara**
### Summary

RAM bundle build time boost using `jest-workers`. Number of workers can be con figured using `--maxWorkers` option, e.g.
```
yarn haul ram-bundle --platform ios --dev false  --max-workers 7
```

### Test plan

Proper test was added to the fixtures 

### To do 
- [x] Think about limiting workers number
- [x] Move setting inferred `maxWorkers` value to `makeConfigFactory`
- [x] Add `--max-workers` documentation
- [x] Add documentation about `--progress` performance implications
- [x] Cleanup `babelWorkerLoader`
- [x] Switch to ESM in `babelWorkerLoader/*`
- [x] Remove commented out code from `preset-*/src/defaultConfig.ts`
- [x] Revert changes to `IndexRamBundle` and FileRamBundle` source map's line generation

### CODE REVIEW

File: packages/haul-ram-bundle-webpack-plugin/package.json [hunk 1]
```diff
@@ -18,5 +18,8 @@
     "webpack": "4.x",
     "webpack-sources": "1.x"
   },
-  "gitHead": "c41becc82dfeafb07b642ebed8f082775647ec3f"
+  "gitHead": "c41becc82dfeafb07b642ebed8f082775647ec3f",
+  "dependencies": {
+    "jest-worker": "^24.9.0"
```
Comment (thymikee) [inline packages/haul-ram-bundle-webpack-plugin/package.json]: FYI, Jest 25 just came out, so you might want to be on the latest :) There is one breaking change to this package, but not sure if it affects you, see https://github.com/facebook/jest/blob/master/CHANGELOG.md

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import fs from 'fs';
+import * as babel from '@babel/core';
+
+import cache from './vendor/cache';
+import transform from './vendor/transform';
+
+const injectCaller = (opts: { [prop: string]: unknown; caller?: any }) => ({
+  ...opts,
+  caller: {
+    name: 'babel-loader',
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
    name: 'babelWorkerLoader',
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import fs from 'fs';
+import * as babel from '@babel/core';
+
+import cache from './vendor/cache';
+import transform from './vendor/transform';
+
+const injectCaller = (opts: { [prop: string]: unknown; caller?: any }) => ({
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
const injectCaller = (opts: Record<string, any>; caller?:  Record<string, any> }) => ({
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+export async function process(
+  this: any,
+  sourceFilename: string,
+  inputSourceMap: string,
+  filename: string,
+  loaderOptions: {
+    sourceMap?: any;
+    sourceMaps?: any;
+    cacheDirectory?: any;
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
    cacheDirectory?: string;
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    cacheIdentifier?: any;
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
    cacheIdentifier?: string;
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    cacheCompression?: any;
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
    cacheCompression?: boolean;
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    inputSourceMap: inputSourceMap || undefined,
+    sourceMaps:
+      loaderOptions.sourceMaps === undefined
+        ? sourceMap
+        : loaderOptions.sourceMaps,
+    sourceFileName: filename,
+  };
+
+  // Remove loader related options
+  delete programmaticOptions.customize;
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  delete programmaticOptions.cacheDirectory;
+  delete programmaticOptions.cacheIdentifier;
+  delete programmaticOptions.cacheCompression;
+  delete programmaticOptions.metadataSubscribers;
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  sourceFilename: string,
+  inputSourceMap: string,
+  filename: string,
+  loaderOptions: {
+    sourceMap?: any;
+    sourceMaps?: any;
+    cacheDirectory?: any;
+    cacheIdentifier?: any;
+    cacheCompression?: any;
+    metadataSubscribers?: any;
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    customize?: any;
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    if (options.sourceMaps === 'inline') {
+      // Babel glitch
+      options.sourceMaps = true;
+    }
+
+    const {
+      cacheDirectory = null,
+      cacheCompression = true,
+      metadataSubscribers = [],
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]:
```suggestion
```

File: packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+          cacheCompression,
+        })
+      : await transform(source, options);
+
+    if (result) {
+      const { code, map, metadata } = result;
+
+      metadataSubscribers.forEach((subscriber: any) => {
+        this[subscriber] && this[subscriber](metadata);
+      });
```
Comment (zamotany) [inline packages/haul-core/src/webpack/loaders/babelWorkerLoader/worker.ts]: We can remove it.

File: docs/CLI Commands.md [hunk 13]
```diff
@@ -101,6 +101,14 @@ Path to use for a source map file, e.g. - `build/index.android.bundle.map`.
 
 Path to the webpack haul config. Defaults to `haul.config.js`.
 
+#### `--progress <string>`
+
+Display bundle compilation progress with different verbosity levels. Accepts `none`, `minimal`, `compact`, `expanded` and `verbose`. Note that logging the compilation progress will increase build time. Defaults to `none` when `--dev` is set to `false`.
+
+#### `--max-workers <number>`
+
+Number of workers used to transpile modules. Defaults to the number of the CPUs - 1. Default is limited to `7` on CI, as the number of CPUs happen to be not correct.
```
Comment (zamotany) [inline docs/CLI Commands.md]:
```suggestion
Number of workers used to transpile modules. Defaults to the number of the CPUs - 1 on local machine, but is limited to `7` on CI, as the number of CPUs can happen to be incorrect.
```

File: docs/CLI Commands.md [hunk 14]
```diff
@@ -153,4 +161,12 @@ Sends reload request to all devices that enabled live reload.
...
\ No newline at end of file
+Port on which the packager server is running. Defaults to `8081`.
+
+#### `--progress <string>`
+
+Display bundle compilation progress with different verbosity levels. Accepts `none`, `minimal`, `compact`, `expanded` and `verbose`. Note that logging the compilation progress will increase build time. Defaults to `none` when you are building in production mode.
+
+#### `--max-workers <number>`
+
+Number of workers used to minify RAM bundle and transpile modules. Defaults to the number of the CPUs - 1. Default is limited to `7` on CI, as the number of CPUs happen to be not correct.
```
Comment (zamotany) [inline docs/CLI Commands.md]:
```suggestion
Number of workers used to transpile modules. Defaults to the number of the CPUs - 1 on local machine, but is limited to `7` on CI, as the number of CPUs can happen to be incorrect.
```

File: docs/Configuration.md [hunk 15]
```diff
@@ -129,6 +129,7 @@ The bundle configuration consist of the following properties:
 - `manifestPath?: string` - Absolute path to manifest for an external (pre-built) bundle, if `dll` is `true`.
 - `assetsPath?: string` - Absolute or relative path pointing to where assets for an external (pre-built) bundle are stored. If relative, it will be joined with `path.dirname(bundlePath)`.
 - `copyBundle?: boolean` - Whether to copy bundle, source map and assets of an external (pre-built) bundle to output directory.
+- `maxWorkers?: number` - Number of workers used to transpile modules by Babel loader. For `ram-bundle` workers are used also for the minification. Defaults to the number of the CPUs - 1. Default is limited to 7 on CI, as the number of CPUs happen to be not correct.
```
Comment (zamotany) [inline docs/Configuration.md]:
```suggestion
- `maxWorkers?: number` - Number of workers used to transpile modules by out Babel loader. For `ram-bundle` workers are used also for the minification. Defaults to the number of the CPUs - 1 on local machine, but is limited to `7` on CI, as the number of CPUs can happen to be incorrect.
```

**Pull request merged**